homeassistant/components/recorder/__init__.py

```diff
@@ -386,9 +393,17 @@ def _commit(work):
                 else:
                     session.add(work)
                 session.commit()
+                if close_session:
+                    session.close()
                 return True
-            except sqlalchemy.exc.OperationalError as e:
-                log_error(e, retry_wait=QUERY_RETRY_WAIT, rollback=True)
+            except sqlalchemy.exc.OperationalError as err:
+                session.rollback()
+                log_error(err, retry_wait=QUERY_RETRY_WAIT)
+            except UnicodeEncodeError as err:
```

Comment:
This happens in #4352. It really is a border case, and I would even recommend catching a general exception, then we cover all bases